File: Src/Icm.ContextConsole.Tests/StreamsInteractorTest.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using NUnit.Framework;

[TestFixture()]
public class StreamsInteractorTest
{

	public static IEnumerable GetAskStringTestCases()
	{
		Func<string, bool> valFunc = (string str) => str == "qwer";
		return {
			new TestCaseData(null, null, null).Returns(null).SetName("Empty list, no validation"),
			new TestCaseData(null, { "2" }, null).Returns("2"),
			new TestCaseData({ "2" }, null, null).Returns("2"),
			new TestCaseData({ "2" }, null, valFunc).Returns(null),
			new TestCaseData({ "qwer" }, null, valFunc).Returns("qwer")
		};
	}

	[Test()]
	[Timeout(2000)]
	[TestCaseSource("GetAskStringTestCases")]
	public string AskStringTest(string[] tokenQueue, string[] input, Func<string, bool> valFunc)
	{
		dynamic inter = BuildInteractor(input);
		if (tokenQueue != null) {
			Queue<string> queue = new Queue<string>(tokenQueue);
			inter.TokenQueue = queue;
		}

		dynamic result = inter.AskString("", valFunc);
		return result;
	}

	public static IEnumerable GetAskOneByKeyTestCases()
	{
		List<Tuple<int, string>> empty = new List<Tuple<int, string>>();
		dynamic list = {
			Tuple.Create(1, "a"),
			Tuple.Create(2, "b"),
			Tuple.Create(3, "c")
		};
		dynamic defaultValue = Tuple.Create(1, "a");
		return {
			new TestCaseData(null, null, empty, null).Throws(typeof(ArgumentException)).SetName("Empty list, no default value"),
			new TestCaseData(null, null, list, null).Returns(null),
			new TestCaseData(null, { "2" }, list, null).Returns(Tuple.Create(2, "b")),
			new TestCaseData({ "2" }, null, list, null).Returns(Tuple.Create(2, "b")),
			new TestCaseData(null, { "4" }, list, null).Throws(typeof(InvalidOperationException)),
			new TestCaseData({ "4" }, null, list, null).Throws(typeof(InvalidOperationException)),
			new TestCaseData(null, null, empty, defaultValue).Throws(typeof(ArgumentException)).SetName("Empty list, with default value"),
			new TestCaseData(null, null, list, defaultValue).Returns(defaultValue),
			new TestCaseData(null, { "2" }, list, defaultValue).Returns(Tuple.Create(2, "b")),
			new TestCaseData({ "2" }, null, list, defaultValue).Returns(Tuple.Create(2, "b")),
			new TestCaseData(null, { "4" }, list, defaultValue).Throws(typeof(InvalidOperationException)),
			new TestCaseData({ "4" }, null, list, defaultValue).Throws(typeof(InvalidOperationException))
		};
	}

	[Test()]
	[Timeout(2000)]
	[TestCaseSource("GetAskOneByKeyTestCases")]
	public Tuple<int, string> AskOneByKeyTest(string[] tokenQueue, string[] input, IEnumerable<Tuple<int, string>> values, Tuple<int, string> defaultValue)
	{
		dynamic inter = BuildInteractor(input);
		if (tokenQueue != null) {
			Queue<string> queue = new Queue<string>(tokenQueue);
			inter.TokenQueue = queue;
		}
		dynamic result = inter.AskOneByKey("", values, tup => tup.Item1.ToString, toString: obj => obj.ToString, defaultValue: defaultValue, selectedList: new List<Tuple<int, string>>());
		return result;
	}

}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================

File: Src2/Icm.ContextConsole.Tests/StreamsInteractorTest.vb
Imports NUnit.Framework

<TestFixture>
Public Class StreamsInteractorTest

    Shared Function GetAskStringTestCases() As IEnumerable
        Dim valFunc As Func(Of String, Boolean) = Function(str As String) str = "qwer"
        Return {
            New TestCaseData(Nothing, Nothing, Nothing).Returns(Nothing).SetName("Empty list, no validation"),
            New TestCaseData(Nothing, {"2"}, Nothing).Returns("2"),
            New TestCaseData({"2"}, Nothing, Nothing).Returns("2"),
            New TestCaseData({"2"}, Nothing, valFunc).Returns(Nothing),
            New TestCaseData({"qwer"}, Nothing, valFunc).Returns("qwer")
        }
    End Function

    <Test()>
    <Timeout(2000)>
    <TestCaseSource("GetAskStringTestCases")>
    Public Function AskStringTest(tokenQueue() As String, input() As String, valFunc As Func(Of String, Boolean)) As String
        Dim inter = BuildInteractor(input)
        If tokenQueue IsNot Nothing Then
            Dim queue As New Queue(Of String)(tokenQueue)
            inter.TokenQueue = queue
        End If

        Dim result = inter.AskString("", valFunc)
        Return result
    End Function

    Shared Function GetAskOneByKeyTestCases() As IEnumerable
        Dim empty As New List(Of Tuple(Of Integer, String))
        Dim list = {Tuple.Create(1, "a"), Tuple.Create(2, "b"), Tuple.Create(3, "c")}
        Dim defaultValue = Tuple.Create(1, "a")
        Return {
            New TestCaseData(Nothing, Nothing, empty, Nothing).Throws(GetType(ArgumentException)).SetName("Empty list, no default value"),
            New TestCaseData(Nothing, Nothing, list, Nothing).Returns(Nothing),
            New TestCaseData(Nothing, {"2"}, list, Nothing).Returns(Tuple.Create(2, "b")),
            New TestCaseData({"2"}, Nothing, list, Nothing).Returns(Tuple.Create(2, "b")),
            New TestCaseData(Nothing, {"4"}, list, Nothing).Throws(GetType(InvalidOperationException)),
            New TestCaseData({"4"}, Nothing, list, Nothing).Throws(GetType(InvalidOperationException)),
            New TestCaseData(Nothing, Nothing, empty, defaultValue).Throws(GetType(ArgumentException)).SetName("Empty list, with default value"),
            New TestCaseData(Nothing, Nothing, list, defaultValue).Returns(defaultValue),
            New TestCaseData(Nothing, {"2"}, list, defaultValue).Returns(Tuple.Create(2, "b")),
            New TestCaseData({"2"}, Nothing, list, defaultValue).Returns(Tuple.Create(2, "b")),
            New TestCaseData(Nothing, {"4"}, list, defaultValue).Throws(GetType(InvalidOperationException)),
            New TestCaseData({"4"}, Nothing, list, defaultValue).Throws(GetType(InvalidOperationException))
        }
    End Function

    <Test()>
    <Timeout(2000)>
    <TestCaseSource("GetAskOneByKeyTestCases")>
    Public Function AskOneByKeyTest(tokenQueue() As String, input() As String, values As IEnumerable(Of Tuple(Of Integer, String)), defaultValue As Tuple(Of Integer, String)) As Tuple(Of Integer, String)
        Dim inter = BuildInteractor(input)
        If tokenQueue IsNot Nothing Then
            Dim queue As New Queue(Of String)(tokenQueue)
            inter.TokenQueue = queue
        End If
        Dim result = inter.AskOneByKey("", values, Function(tup) tup.Item1.ToString, toString:=Function(obj) obj.ToString, defaultValue:=defaultValue, selectedList:=New List(Of Tuple(Of Integer, String)))
        Return result
    End Function

End Class